docs/tutorials/automatic-differentiation.md

```diff
@@ -1,9 +1,220 @@
+
+## Differentiating with respect to nested lists, tuples, and dicts
+
+Differentiating with respect to standard Python containers just works, so use tuples, lists, and dicts (and arbitrary nesting) however you like.
+
+Continuing the previous example:
+
+```{code-cell}
+def loss2(params_dict):
+    preds = predict(params_dict['W'], params_dict['b'], inputs)
+    label_probs = preds * targets + (1 - preds) * (1 - targets)
+    return -jnp.sum(jnp.log(label_probs))
+
+print(grad(loss2)({'W': W, 'b': b}))
+```
+
+You can [register your own container types](https://jax.readthedocs.io/en/latest/pytrees.html#extending-pytrees) to work with not just {func}`jax.grad` but other JAX transformations ({func}`jax.jit`, {func}`jax.vmap`, and so on).
+
+## Evaluate a function and its gradient using `value_and_grad`
+
+Another convenient function is {func}`jax.value_and_grad` for efficiently computing both a function's value as well as its gradient's value.
+
+Continuing the previous examples:
+
+```{code-cell}
+loss_value, Wb_grad = jax.value_and_grad(loss, (0, 1))(W, b)
+print('loss value', loss_value)
+print('loss value', loss(W, b))
+```
+
+## Checking against numerical differences
+
+A great thing about derivatives is that they're straightforward to check with finite differences.
+
+Continuing the previous examples:
+
+```{code-cell}
+# Set a step size for finite differences calculations
+eps = 1e-4
+
+# Check b_grad with scalar finite differences
+b_grad_numerical = (loss(W, b + eps / 2.) - loss(W, b - eps / 2.)) / eps
+print('b_grad_numerical', b_grad_numerical)
+print('b_grad_autodiff', grad(loss, 1)(W, b))
+
+# Check W_grad with finite differences in a random direction
+key, subkey = jax.random.split(key)
+vec = jax.random.normal(subkey, W.shape)
+unitvec = vec / jnp.sqrt(jnp.vdot(vec, vec))
+W_grad_numerical = (loss(W + eps / 2. * unitvec, b) - loss(W - eps / 2. * unitvec, b)) / eps
+print('W_dirderiv_numerical', W_grad_numerical)
+print('W_dirderiv_autodiff', jnp.vdot(grad(loss)(W, b), unitvec))
+```
+
+JAX provides a simple convenience function that does essentially the same thing, but checks up to any order of differentiation that you like:
+
+```{code-cell}
+from jax.test_util import check_grads
+
+check_grads(loss, (W, b), order=2)  # check up to 2nd order derivatives
+```
 
-```{note}
-This is a placeholder for a section in the new {ref}`jax-tutorials`.
+## Next steps
 
-For the time being, you may find some related content in the old documentation:
-- {doc}`../jax-101/04-advanced-autodiff`
-```
\ No newline at end of file
+The {ref}`advanced-autodiff` tutorial provides more advanced and detailed explanations of how the ideas covered in this document are implemented in the JAX backend. Some features, such as (defining [custom derivatives](https://jax.readthedocs.io/en/latest/Custom_derivative_rules_for_Python_code.html), depend on understanding advanced automatic differentiation, so do check out the tutorial if you are interested.
```

Comment:
We should use a `{ref}` here as well, or maybe ``` {doc}`../Custom_derivative_rules_for_Python_code` ``` since the target is a notebook and we may not want to add markdown reference targets to it.

Reply:
:+1: 

```
{ref}`defining-custom-derivative-rules`
```

which will be in

```
{ref}`advanced-autodiff`
```